File: apps/vis-companion/src/lib/activity.test.ts
/** The Activity projection against `internal.activity.core`'s own bounded snapshot. */
import { describe, expect, it, vi } from 'vitest';
import {
  activityCopyText,
  activityHistoryCopyText,
  mergeActivity,
  argumentGroups,
  operationGroups,
  activityProjectionFromWire,
  ACTIVITY_PRESENTERS,
  ACTIVITY_SIGNALS,
  ACTIVITY_STATES,
  ACTIVITY_TEXT_FORMATS,
} from './activity';
import contract from '../../../../packages/vis-contract/resources/vis-contract/schema/activity.json';
import cases from '../../../../packages/vis-contract/resources/vis-contract/fixtures/activity-cases.json';
import groupingCases from '../../../../packages/vis-contract/resources/vis-contract/fixtures/activity-groups.json';
import argumentCases from '../../../../packages/vis-contract/resources/vis-contract/fixtures/activity-arguments.json';
import copyCases from '../../../../packages/vis-contract/resources/vis-contract/fixtures/activity-copy.json';
import {
  activityHistoryPage,
  groupedActivityHistoryPage,
  GROUPED_ACTIVITY_HISTORY_IDS,
} from '../dev/activity-history';

for (const sample of copyCases) {
  it(`portable Activity copy: ${sample.name}`, () => {
    const projection = activityProjectionFromWire(sample.projection);
    expect(activityCopyText(projection!)).toBe(sample.text);
  });
}
for (const sample of argumentCases) {
  it(`portable argument grouping: ${sample.name}`, () => {
    const projection = activityProjectionFromWire(sample.projection);
    expect(projection).toEqual(sample.projection);
    expect(
      argumentGroups(projection!.rows).map(({ id, rows }) => ({
        id,
        rows: rows.map((row) => row.id),
      })),
    ).toEqual(sample.groups);
  });
}
// Regression #201: shared fixtures include extension headlines and fallback order.
for (const sample of groupingCases) {
  it(`portable grouping: ${sample.name}`, () => {
    const projection = activityProjectionFromWire(sample.projection);
    expect(
      operationGroups(projection!.rows).map(({ id, label, rows }) => ({
        id,
        label,
        rows: rows.map((row) => row.id),
      })),
    ).toEqual(sample.groups);
    expect(projection).toEqual(sample.projection);
  });
}
const activityProjection = (
  state: 'running' | 'succeeded' | 'failed' | 'cancelled' = 'running',
) => ({
  state,
  counts: {
    running: state === 'running' ? 1 : 0,
    succeeded: state === 'succeeded' ? 1 : 0,
    failed: state === 'failed' ? 1 : 0,
    cancelled: state === 'cancelled' ? 1 : 0,
  },
  rows: [
    {
      id: 'call-1',
      sequence: 1,
      operation: 'suite',
      presenter: 'tests',
      signal: 'verification',
      state,
      summary: 'suite',
      resources: [],
      evidence: [{ kind: 'arguments', text: 'suite' }],
    },
  ],
  omitted: { rows: 0, by_classification: {} },
});

describe('canonical Activity admission across SDK, engine and surfaces', () => {
  it('uses the canonical vocabulary', () => {
    expect(ACTIVITY_PRESENTERS).toEqual(contract.$defs.presenter.enum);
    expect(ACTIVITY_SIGNALS).toEqual(contract.$defs.signal.enum);
    expect(ACTIVITY_STATES).toEqual(contract.$defs.state.enum);
    expect(ACTIVITY_TEXT_FORMATS).toEqual(contract.$defs.text_format.enum);
  });
  for (const sample of cases) {
    it(sample.name, () => {
      expect(activityProjectionFromWire(sample.projection)).toEqual(
        sample.valid ? sample.projection : null,
      );
    });
  }
});

describe("one form's Activity read off the wire", () => {
  // Protocol 7 took Activity off the Live View rail: it is no longer a classified
  // view with a projection hanging off it. Protocol 9 carries every revision on
  // `block.activity`; only the settled replacement is durable.
  it('reads a bare projection off the wire', () => {
    expect(activityProjectionFromWire(activityProjection())).toMatchObject({
      state: 'running',
    });
  });

  it('reads structured diff evidence and rejects incomplete lines', () => {
    const projection = activityProjection();
    const diff = {
      kind: 'diff',
      text: 'fixture.clj',
      lines: [
        { kind: 'hunk', text: '@@ -1 +1 @@' },
        { kind: 'deletion', text: '[REDACTED]', is_redacted: true },
        { kind: 'addition', text: 'after' },
      ],
      additions: 0,
      deletions: 0,
      modifications: 1,
      is_truncated: true,
      is_redacted: true,
    };
    const withDiff = {
      ...projection,
      rows: [{ ...projection.rows[0], evidence: [diff] }],
    };
    expect(activityProjectionFromWire(withDiff)?.rows[0].evidence[0]).toEqual(diff);
    expect(
      activityProjectionFromWire({
        ...withDiff,
        rows: [
          {
            ...withDiff.rows[0],
            evidence: [{ ...diff, lines: [{ kind: 'addition' }] }],
          },
        ],
      }),
    ).toBeNull();
  });

  it('rejects a missing, malformed, or retired Activity projection', () => {
    expect(activityProjectionFromWire(undefined)).toBeNull();
    expect(
      activityProjectionFromWire({
        ...activityProjection(),
        rows: [{ id: 'broken' }],
      }),
    ).toBeNull();
    // Protocol 8 dropped both: a payload still wearing either came from a gateway
    // the compatibility gate should already have refused.
    expect(
      activityProjectionFromWire({
        ...activityProjection(),
        schema_version: 1,
      }),
    ).toBeNull();
    expect(
      activityProjectionFromWire({
        ...activityProjection(),
        anchor: { iteration: 1, form_index: 0 },
      }),
    ).toBeNull();
  });

  it('rejects projections outside the one canonical closed shape', () => {
    const projection = activityProjection();
    const row = projection.rows[0];

    expect(activityProjectionFromWire({ ...projection, extra: true })).toBeNull();
    expect(
      activityProjectionFromWire({
        ...projection,
        counts: { ...projection.counts, total: 1 },
      }),
    ).toBeNull();
    expect(
      activityProjectionFromWire({
        ...projection,
        omitted: { ...projection.omitted, total: 0 },
      }),
    ).toBeNull();
    expect(activityProjectionFromWire({ ...projection, rows: [row, { ...row }] })).toBeNull();
    expect(
      activityProjectionFromWire({
        ...projection,
        rows: [{ ...row, duration_ms: 1.5 }],
      }),
    ).toBeNull();
    expect(
      activityProjectionFromWire({
        ...projection,
        rows: [{ ...row, resources: [{ type: 'file', id: 'a.clj', extra: true }] }],
      }),
    ).toBeNull();
  });

  it('carries a per-field text format and refuses one it does not know', () => {
    const projection = activityProjection();
    const row = projection.rows[0];
    const marked = {
      ...projection,
      rows: [
        {
          ...row,
          summary_format: 'inline',
          result_summary: 'wrote `a.clj`',
          result_format: 'markdown',
        },
      ],
    };

    expect(activityProjectionFromWire(marked)?.rows[0].summary_format).toBe('inline');
    expect(activityProjectionFromWire(marked)?.rows[0].result_format).toBe('markdown');
    // A row with no flag stays plain text; the app never guesses a format.
    expect(activityProjectionFromWire(projection)?.rows[0].summary_format).toBeUndefined();
    expect(
      activityProjectionFromWire({
        ...projection,
        rows: [{ ...row, summary_format: 'html' }],
      }),
    ).toBeNull();
    expect(
      activityProjectionFromWire({
        ...projection,
        rows: [{ ...row, result_format: true }],
      }),
    ).toBeNull();
  });
});

// #212: a bounded page is not a retention cap.
it('admits a durable history window beyond the former receipt size', () => {
  const page = activityProjection('succeeded');
  const row = page.rows[0];
  const rows = Array.from({ length: 24 }, (_, index) => ({
    ...row,
    id: `call-${index}`,
    sequence: index + 1,
    presentation: {
      headline: 'Read',
      summary: '',
      content: [{ type: 'code', language: 'text', text: 'x'.repeat(4000) }],
    },
  }));
  const history = {
    id: '12345678-1234-1234-1234-123456789012',
    revision: 1,
    total: 160,
    after: 0,
    next_after: 24,
  };
  expect(activityProjectionFromWire({ ...page, rows, history })).toMatchObject({
    history,
    rows,
  });
  for (const invalid of [
    { ...history, next_after: 0 },
    { ...history, total: -1 },
    { ...history, id: '' },
  ]) {
    expect(activityProjectionFromWire({ ...page, rows, history: invalid })).toBeNull();
  }
  expect(
    activityProjectionFromWire({
      ...page,
      history,
      rows: Array.from({ length: 33 }, (_, i) => ({ ...row, id: `over-${i}`, sequence: i + 1 })),
    }),
  ).toBeNull();
  // Inline history has no total row cap; only history-bearing transport windows do.
  expect(
    activityProjectionFromWire({
      ...page,
      rows: Array.from({ length: 160 }, (_, i) => ({ ...row, id: `inline-${i}`, sequence: i + 1 })),
    })?.rows,
  ).toHaveLength(160);
  const children = Array.from({ length: 32 }, (_, i) => ({
    ...row,
    id: `nested-${i}`,
    sequence: i + 1,
  }));
  const grouped = {
    ...page,
    history,
    rows: [{ ...row, id: 'group', operation: 'shell', children }],
  };
  expect(activityProjectionFromWire(grouped)?.rows[0].children).toHaveLength(32);
  expect(
    activityProjectionFromWire({ ...grouped, rows: [{ ...grouped.rows[0], id: children[0].id }] }),
  ).toBeNull();
});

// #218: complete result content must survive admission and copying, not only the headline.
describe('complete Activity presentation content', () => {
  const body = `${'x'.repeat(1_100_000)} final-code-detail`;
  const cell = `${'y'.repeat(300)} final-cell-detail`;
  const columns = Array.from({ length: 17 }, (_, i) => `Field ${i}`);
  const rows = Array.from({ length: 201 }, (_, i) =>
    columns.map((_, j) => (i === 200 && j === 16 ? cell : `${i}-${j}`)),
  );
  const presentation = {
    headline: 'Find builds for review',
    summary: 'Complete build results',
    content: [
      { type: 'code', language: 'text', text: body },
      { type: 'table', columns, rows },
      ...Array.from({ length: 33 }, (_, i) => ({ type: 'text', text: `Detail ${i}` })),
    ],
    sections: Array.from({ length: 9 }, (_, i) => ({
      headline: `Build ${i}`,
      summary: 'Completed',
      content: [{ type: 'text', text: `section-${i}-final-detail` }],
    })),
  };
  const page = activityProjection('succeeded');
  const row = { ...page.rows[0], presentation };
  const history = {
    id: '12345678-1234-1234-1234-123456789012',
    revision: 1,
    total: 2,
    after: 0,
    next_after: 1,
  };

  it('retains all large blocks, table cells and sections through parsing and copying', () => {
    const parsed = activityProjectionFromWire({ ...page, rows: [row] });
    expect(parsed!.rows[0].presentation).toEqual(presentation);
    const copied = activityCopyText(parsed!);
    expect(copied).toContain(body);
    expect(copied).toContain(cell);
    expect(copied).toContain('Detail 32');
    expect(copied).toContain('section-8-final-detail');
    expect(copied).not.toContain('Details truncated');
  });

  it('admits one oversized invocation on a history page without losing its cursor', () => {
    const parsed = activityProjectionFromWire({ ...page, rows: [row], history });
    expect(parsed!.history).toEqual(history);
    expect(parsed!.rows[0].presentation).toEqual(presentation);
    expect(
      activityProjectionFromWire({
        ...page,
        rows: [{ ...page.rows[0], id: 'group', children: [row] }],
        history,
      })?.rows[0].children?.[0].presentation,
    ).toEqual(presentation);
  });

  it('still requires multiple invocations to fit the history page byte budget', () => {
    expect(
      activityProjectionFromWire({
        ...page,
        rows: [row, { ...row, id: 'call-2', sequence: 2 }],
        history: { ...history, next_after: null },
      }),
    ).toBeNull();
    expect(
      activityProjectionFromWire({
        ...page,
        rows: [
          { ...page.rows[0], id: 'group', children: [row, { ...row, id: 'call-2', sequence: 2 }] },
        ],
        history: { ...history, next_after: null },
      }),
    ).toBeNull();
  });

  it('still rejects malformed content rather than treating it as complete evidence', () => {
    for (const content of [
      [{ type: 'code', text: 1 }],
      [{ type: 'table', columns: ['Name', 'Result'], rows: [['missing result']] }],
      [{ type: 'progress', label: 'Builds', value: 3, total: 2 }],
    ]) {
      expect(
        activityProjectionFromWire({
          ...page,
          rows: [{ ...row, presentation: { ...presentation, content } }],
        }),
      ).toBeNull();
    }
  });
});

describe('complete retained Activity copy', () => {
  it('copies every page from the beginning without mutating the visible snapshot', async () => {
    const snapshot = activityHistoryPage(96);
    const original = structuredClone(snapshot);
    const signal = new AbortController().signal;
    const load = vi.fn(async (_id: string, after: number) => activityHistoryPage(after));
    const text = await activityHistoryCopyText([snapshot], load, signal);
    expect(load.mock.calls.map(([, after]) => after)).toEqual([0, 32, 64, 96, 128]);
    expect(text.match(/^operation \d+ \[/gm)).toHaveLength(160);
    expect(text).toContain('retained-1\n');
    expect(text).toContain('retained-160\n');
    expect(text).toContain('operation 159 [failed]');
    expect(text).toContain('operation 160 [cancelled]');
    expect(text).not.toContain(snapshot.history!.id);
    expect(text).not.toContain('op-160');
    expect(snapshot).toEqual(original);
  });

  it('preserves mixed source order and repeated calls without reloading complete histories', async () => {
    const complete = groupedActivityHistoryPage(GROUPED_ACTIVITY_HISTORY_IDS[0]);
    const inline = structuredClone(complete);
    delete inline.history;
    const tail = groupedActivityHistoryPage(GROUPED_ACTIVITY_HISTORY_IDS[2], 2);
    const load = vi.fn(async (id: string, after: number) => groupedActivityHistoryPage(id, after));
    const full = groupedActivityHistoryPage(tail.history!.id);
    full.rows.push(...tail.rows);
    const text = await activityHistoryCopyText(
      [complete, inline, tail],
      load,
      new AbortController().signal,
    );
    expect(load.mock.calls).toEqual([
      [tail.history!.id, 0, '', expect.any(AbortSignal)],
      [tail.history!.id, 2, '', expect.any(AbortSignal)],
    ]);
    expect(text).toBe(activityCopyText(mergeActivity([complete, inline, full])));
    expect(text.match(/Read result 1-1/g)).toHaveLength(2);
  });

  it('copies nested invocation content in complete snapshots without a loader call', async () => {
    const snapshot = groupedActivityHistoryPage(GROUPED_ACTIVITY_HISTORY_IDS[0]);
    snapshot.rows = [{ ...snapshot.rows[0], children: snapshot.rows }];
    const load = vi.fn();
    expect(await activityHistoryCopyText([snapshot], load, new AbortController().signal)).toBe(
      activityCopyText(snapshot),
    );
    expect(load).not.toHaveBeenCalled();
  });

  it.each(['id', 'after', 'revision', 'total', 'cursor', 'missing', 'empty', 'empty tail'])(
    'rejects a changed or invalid %s rather than returning a partial copy',
    async (field) => {
      const load = vi.fn(async (_id: string, after: number) => {
        const result = activityHistoryPage(after);
        if (after === 32) {
          if (field === 'id') result.history!.id = GROUPED_ACTIVITY_HISTORY_IDS[0];
          if (field === 'after') result.history!.after = 0;
          if (field === 'revision') result.history!.revision++;
          if (field === 'total') result.history!.total++;
          if (field === 'cursor') result.history!.next_after = after;
          if (field === 'missing') delete result.history;
          if (field === 'empty') result.rows = [];
          if (field === 'empty tail') {
            result.rows = [];
            result.history!.next_after = null;
          }
        }
        return result;
      });
      await expect(
        activityHistoryCopyText([activityHistoryPage()], load, new AbortController().signal),
      ).rejects.toThrow(/Activity changed/);
      expect(load).toHaveBeenCalledTimes(2);
    },
  );

  it('rejects a revision change even on the first reloaded page', async () => {
    const load = vi.fn(async () => {
      const result = activityHistoryPage();
      result.history!.revision++;
      return result;
    });
    await expect(
      activityHistoryCopyText([activityHistoryPage()], load, new AbortController().signal),
    ).rejects.toThrow(/Activity changed/);
  });

  it('propagates retrieval failures without returning the pages already loaded', async () => {
    const load = vi.fn(async (_id: string, after: number) => {
      if (after) throw new Error('Reconnect to copy activity.');
      return activityHistoryPage();
    });
    await expect(
      activityHistoryCopyText([activityHistoryPage()], load, new AbortController().signal),
    ).rejects.toThrow('Reconnect to copy activity.');
  });

  it.each([false, true])(
    'honors cancellation before or during retrieval (during: %s)',
    async (during) => {
      const controller = new AbortController();
      const reason = new Error('Copy cancelled');
      const load = vi.fn(async () => {
        controller.abort(reason);
        return activityHistoryPage();
      });
      if (!during) controller.abort(reason);
      await expect(
        activityHistoryCopyText([activityHistoryPage()], load, controller.signal),
      ).rejects.toBe(reason);
      expect(load).toHaveBeenCalledTimes(during ? 1 : 0);
    },
  );
});

File: test/com/blockether/vis/contract/activity_test.clj
(ns com.blockether.vis.contract.activity-test
  "The same Activity receipt is consumed by Python, Companion and TUI."
  (:require [clojure.data.json :as json]
            [com.blockether.vis.contract.document :as document]
            [clojure.java.io :as io]
            [com.blockether.vis.contract.activity :as activity]
            [com.blockether.vis.contract.wire :as wire]
            [lazytest.experimental.interfaces.clojure-test :refer [deftest is]]))

(deftest shared-activity-copy-test
  (doseq [{:strs [name projection text]}
          (json/read-str (slurp (io/resource "vis-contract/fixtures/activity-copy.json")))]
    (let [receipt (activity/from-wire projection)]
      (is (some? receipt) name)
      (is (= text (activity/copy-text receipt)) name))))

(deftest portable-activity-contract-test
  (let [fixture (json/read-str (slurp (io/resource "vis-contract/fixtures/activity.json")))]
    (is (nil? (io/resource "vis-contract/activity.json")))
    (is (document/valid? "activity" fixture))
    (is (document/valid-json? "activity" "projection" fixture))
    (doseq [old-key ["anchor" "schema_version" "view_id"]]
      (is (not (document/valid-json? "activity" "projection" (assoc fixture old-key 1)))))
    (doseq [show-start [true false]]
      (is (document/valid-json?
            "activity"
            "declaration"
            {"presenter" "tests" "label" "Check components" "show_start" show-start})))
    (doseq [show-start [nil 0 1 "false"]]
      (is (not (document/valid-json? "activity"
                                     "declaration"
                                     {"presenter" "tests" "show_start" show-start}))))
    (is (not (document/valid-json? "activity"
                                   "declaration"
                                   {"presenter" "tests" "state" "succeeded"})))))

(deftest persistent-handle-id-contract-test
  (let [fixture
        (json/read-str (slurp (io/resource "vis-contract/fixtures/activity.json")))

        view
        {"headline" "Checking visual differences" "summary" "Done" "content" []}]

    (doseq [id ["comparison-1" (apply str (repeat 256 "é"))]]
      (is (activity/valid-handle-id? id))
      (is (activity/valid-presentation? (assoc view "handle_id" id)))
      (is (activity/valid-projection? (assoc-in fixture ["rows" 0 "handle_id"] id))))
    (doseq [id ["" "  " (apply str (repeat 257 "é")) (apply str (repeat 171 "界")) "line\nbreak"
                (str "line" (char 8232) "break") true]]
      (is (not (activity/valid-handle-id? id)))
      (is (not (activity/valid-presentation? (assoc view "handle_id" id))))
      (is (not (activity/valid-projection? (assoc-in fixture ["rows" 0 "handle_id"] id)))))))

(deftest shared-activity-admission-test
  (doseq [{:strs [name valid projection]}
          (json/read-str (slurp (io/resource "vis-contract/fixtures/activity-cases.json")))]
    (is (= valid (boolean (activity/from-wire projection))) name)
    (when valid (is (= projection (wire/->wire (activity/from-wire projection))) name))))

(deftest shared-activity-digest-admission-test
  (doseq [{:strs [name valid digest]}
          (json/read-str (slurp (io/resource "vis-contract/fixtures/activity-digest.json")))]
    (is (= valid (boolean (activity/digest-from-wire digest))) name)
    (when valid (is (= digest (wire/->wire (activity/digest-from-wire digest))) name))))

(deftest shared-operation-groups-test
  ;; Regression #201: extension labels must agree with Companion without changing receipts.
  (doseq [{:strs [name projection groups]}
          (json/read-str (slurp (io/resource "vis-contract/fixtures/activity-groups.json")))]
    (let [receipt (activity/from-wire projection)
          group-rows (ns-resolve 'com.blockether.vis.contract.activity 'operation-groups)]

      (is (some? receipt) name)
      (is (= projection (wire/->wire receipt)) name)
      (is (some? group-rows) "The contract owns chronological grouping")
      (when group-rows
        (is (= groups
               (mapv (fn [{:keys [id label rows]}]
                       {"id" id "label" label "rows" (mapv :id rows)})
                     (group-rows (:rows receipt))))
            name)))))

(deftest shared-argument-groups-test
  (doseq [{:strs [name projection groups]}
          (json/read-str (slurp (io/resource "vis-contract/fixtures/activity-arguments.json")))]
    (let [receipt (activity/from-wire projection)
          group-rows (ns-resolve 'com.blockether.vis.contract.activity 'argument-groups)]

      (is (some? receipt) name)
      (is (= projection (wire/->wire receipt)) name)
      (is (some? group-rows))
      (when group-rows
        (is (= groups
               (mapv (fn [{:keys [id rows]}]
                       {"id" id "rows" (mapv :id rows)})
                     (group-rows (:rows receipt))))
            name)))))

(deftest complete-presentation-content-test
  ;; Regression #218: retain full result bodies rather than failing size admission.
  (let [text
        (apply str (repeat 20000 "界"))

        block
        {"type" "code" "text" text}

        table
        {"type" "table"
         "columns" (vec (repeat 17 (apply str (repeat 300 "C"))))
         "rows" (vec (repeat 201 (vec (repeat 17 (apply str (repeat 300 "x"))))))}

        content
        (conj (vec (repeat 40 block)) table)

        section
        {"headline" "Results" "summary" "Complete" "content" content}

        presentation
        (assoc section "sections" (vec (repeat 9 section)))]

    (is (activity/valid-presentation? presentation))
    (is (not (activity/valid-presentation? (assoc presentation "headline" "two\nlines"))))
    (is (not (activity/valid-presentation?
               (assoc presentation
                 "content" [{"type" "table" "columns" ["A"] "rows" [["A" "B"]]}]))))))

(deftest table-row-paths-test
  ;; BLO-172: a table row may NAME the file it lists, so a reader can open it from
  ;; the row. One path per row or none at all — a shorter list presses the wrong row.
  (let [table
        {"type" "table" "columns" ["Name"] "rows" [["a.clj"] ["b.clj"]]}

        presentation
        {"headline" "Listed directory"
         "summary" "2 files"
         "content" [(assoc table "paths" ["/w/a.clj" ""])]}]

    (is (activity/valid-presentation? presentation))
    (is (activity/valid-presentation? (assoc presentation "content" [table])))
    (is (not (activity/valid-presentation? (assoc presentation
                                             "content" [(assoc table "paths" ["/w/a.clj"])]))))))

(deftest presentation-summary-format-test
  ;; Regression #254: only explicitly marked summaries opt into inline Markdown.
  (let [summary
        "Issues: [#252](https://github.com/Blockether/vis/issues/252)"

        section
        {"headline" "Matches" "summary" summary "content" []}]

    (is (activity/valid-presentation? section))
    (doseq [format ["inline" "markdown"]]
      (let [formatted (assoc section "summary_format" format)
            presentation (assoc formatted "sections" [formatted])]

        (is (activity/valid-presentation? presentation))
        (is (not (activity/valid-presentation? (assoc presentation "summary" "two\nlines")))))
      (is (not (activity/valid-presentation? (assoc section
                                               "summary_format" format
                                               "summary" (apply str (repeat 257 "é")))))))
    (doseq [format [nil "" "html" true 7 []]]
      (is (not (activity/valid-presentation? (assoc section "summary_format" format)))))))

(deftest single-oversized-invocation-page-test
  ;; Regression #218: page bytes are a target; one complete invocation always fits.
  (let [fixture
        (json/read-str (slurp (io/resource "vis-contract/fixtures/activity.json")))

        row
        (assoc (first (get fixture "rows"))
          "presentation" {"headline" "Result"
                          "summary" "Complete"
                          "content" [{"type" "code" "text" (apply str (repeat 400000 "界"))}]})

        page
        (assoc fixture
          "rows" [row]
          "history" {"id" "00000000-0000-4000-8000-000000000218"
                     "revision" 1
                     "total" 1
                     "after" 0
                     "next_after" nil})]

    (is (activity/valid-projection? page))
    (is (= page (wire/->wire (activity/from-wire page))))
    (is (not (activity/valid-projection? (update page
                                                 "rows"
                                                 conj
                                                 (assoc row
                                                   "id" "second"
                                                   "sequence" 2)))))))
